src/conn/pool/mod.rs

```diff
@@ -338,6 +338,75 @@ impl Pool {
         self.poll_new_conn_inner(false)
     }
 
+    /// Try to acquire a new connection under a spinlock checking that we don't
+    /// violate `exist < self.pool_constraints.max()`.
+    fn spin_acquire_conn(&mut self) -> Option<Result<Async<GetConn>>> {
+        let mut exist = self.inner.exist.load(atomic::Ordering::Acquire);
+
+        // Try to atomically swap the exist value while checking our condition.
+        loop {
+            if exist >= self.pool_constraints.max() {
+                return None;
+            }
+
+            let swapped =
+                self.inner
+                    .exist
+                    .compare_and_swap(exist, exist + 1, atomic::Ordering::AcqRel);
+
+            if exist == swapped {
+                break;
+            }
+
+            exist = swapped;
+        }
+
+        // we're allowed to make a new connection
+
+        // note, however, that there is a race here:
+        // imagine that Pool::disconnect was _just_ called. that is, after we checked at
+        // the start of this method. the Recycler checks .exist right _before_ we increment
+        // it, and notices that it is 0, and thus believes it is allowed to exit. if we
+        // continue to make a new connection here, that connection would not have a way to
+        // be dropped as part of the pool.
+        //
+        // so, we check .close again here (after the increment). if it is now true, we know
+        // that the Recycler may have exited, and we give up. if it is false, we _know_
+        // that the Recyler _must_ see our +1 before it decides to exit.
+        if self.inner.close.load(atomic::Ordering::Acquire) {
+            self.inner.exist.fetch_sub(1, atomic::Ordering::AcqRel);
+            // make sure we notify the Recycler in case it was waiting for our +1
+            self.drop
+                .try_send(None)
+                .expect("recycler is active as long as any Pool is");
+            return Some(Err(Error::Driver(DriverError::PoolDisconnected)));
+        }
+
+        // We may be allowed to create a new recycler!
```

Comment:
Ah, this isn't sometimes creating a _new_ recycler, it should (in theory) only happen exactly once to create _the_ recycler. May be worth updating the comment to reflect this.

Reply:
Yeah :). Separate PR since I'll be closing this. Which reminds me. This could probably use [`ArcSwap`](https://docs.rs/arc-swap).